api/client/src/main/java/org/projectnessie/client/rest/ResponseCheckFilter.java

```diff
@@ -132,4 +140,43 @@ private static NessieError decodeErrorObject(
     }
     return error;
   }
+
+  /**
+   * Captures the first 2kB of the input in case the response is not parsable, to provide a better
+   * error message to the user.
+   */
+  static final class CapturingInputStream extends FilterInputStream {
+    static final int CAPTURE_LEN = 2048;
+    private final byte[] capture = new byte[CAPTURE_LEN];
+    private int captured;
+
+    CapturingInputStream(InputStream delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+      int rd = super.read(b, off, len);
+      int captureRemain = capture.length - captured;
+      if (rd > 0 && captureRemain > 0) {
+        int copy = Math.min(rd, captureRemain);
+        System.arraycopy(b, off, capture, captured, copy);
+        captured += copy;
+      }
+      return rd;
+    }
+
+    @Override
+    public int read() throws IOException {
+      int rd = super.read();
+      if (rd >= 0 && captured < capture.length) {
+        capture[captured++] = (byte) rd;
+      }
+      return rd;
+    }
+
+    public String captured() {
+      return new String(capture, 0, captured, UTF_8);
```

Comment:
Can we get encoding from the HTTP response?

Reply:
Why? Unparseable characters are represented as `?` - so IMO that would be overkill (mapping character-set names) and become error-prone.

Reply:
fair enough.